File: ScheduleView/Database/SL_VB/WithDB/ScheduleViewRepository.vb
Imports ScheduleViewDB.Web

Public Class ScheduleViewRepository

    Private Shared m_context As SVDomainContext
    Public Shared ReadOnly Property Context() As SVDomainContext
        Get
            If m_context Is Nothing Then
                m_context = New SVDomainContext()
            End If

            Return m_context
        End Get
    End Property

    Shared Sub New()
    End Sub

    Public Shared Function SaveData(ByVal action As Action) As Boolean
        If ScheduleViewRepository.Context.HasChanges AndAlso (Not ScheduleViewRepository.Context.IsSubmitting) Then
            Try
                ScheduleViewRepository.Context.SubmitChanges(AddressOf OnSubmitChangesCompleted, action)
                Return True
            Catch e1 As System.Exception
                Return False
            End Try
        End If

        If action IsNot Nothing Then
            action()
        End If

        Return False
    End Function

    Private Shared Sub OnSubmitChangesCompleted(ByVal submitOperation As ServiceModel.DomainServices.Client.SubmitOperation)
        If Not submitOperation.HasError Then
            Dim action As Action = TryCast(submitOperation.UserState, Action)
            If action IsNot Nothing Then
                action()
            End If
        End If
    End Sub

End Class

File: ScheduleView/Database/SL_CS/WithDB/ScheduleViewRepository.cs
using System;
using System.ServiceModel.DomainServices.Client;
using ScheduleViewDB.Web;

namespace ScheduleViewDB
{
	public class ScheduleViewRepository
	{
		private static ScheduleViewDomainContext context;

		static ScheduleViewRepository() { }

		public static ScheduleViewDomainContext Context
		{
			get
			{
				if (context == null)
				{
					context = new ScheduleViewDomainContext();
				}

				return context;
			}
		}

		public static bool SaveData(Action action)
		{
            if (ScheduleViewRepository.Context.HasChanges && !ScheduleViewRepository.Context.IsSubmitting)
            {
                try
                {
                    ScheduleViewRepository.Context.SubmitChanges(OnSubmitChangesCompleted, action);
                    return true;
                }
                catch (System.Exception)
                {
                    return false;
                }
            }

            if (action != null)
            {
                action();
            }

            return false;
		}

		private static void OnSubmitChangesCompleted(SubmitOperation submitOperation)
		{
			var action = submitOperation.UserState as Action;

			if (action != null)
			{
				action();
			}

			if (submitOperation.HasError)
			{				
				if (submitOperation.Error is DomainOperationException)
				{
					DomainOperationException exception = (DomainOperationException)submitOperation.Error;

					if (exception.Status == OperationErrorStatus.Conflicts)
					{
						foreach (var item in submitOperation.EntitiesInError)
						{
							if (item.EntityConflict.IsDeleted)
							{
								foreach (var entitySet in ScheduleViewRepository.Context.EntityContainer.EntitySets)
								{
									if (entitySet.EntityType == item.GetType())
									{
										entitySet.Detach(item);
										entitySet.Attach(item);
										entitySet.Remove(item);
										ScheduleViewRepository.Context.SubmitChanges();
									}
								}
							}
							else
							{
								item.EntityConflict.Resolve();
							}
						}
					}
					submitOperation.MarkErrorAsHandled();
				}
			}
		}
	}
}
